api/apps/geoprocessing/src/modules/cost-surface/adapters/scenario/typeorm-scenario-cost-surface.ts

```diff
@@ -0,0 +1,66 @@
+import { CHUNK_SIZE_FOR_BATCH_GEODB_OPERATIONS } from '@marxan-geoprocessing/utils/chunk-size-for-batch-geodb-operations';
+import { Injectable } from '@nestjs/common';
+import { InjectEntityManager } from '@nestjs/typeorm';
+import { chunk } from 'lodash';
+import { EntityManager } from 'typeorm';
+import { CostSurfacePuDataEntity } from '@marxan/cost-surfaces';
+import { geoprocessingConnections } from '@marxan-geoprocessing/ormconfig';
+import { ScenarioCostSurfacePersistencePort } from '@marxan-geoprocessing/modules/cost-surface/ports/persistence/scenario-cost-surface-persistence.port';
+import {
+  ScenariosPuCostDataGeo,
+  ScenariosPuPaDataGeo,
+} from '@marxan/scenarios-planning-unit';
+
+@Injectable()
+export class TypeormScenarioCostSurface
+  implements ScenarioCostSurfacePersistencePort {
+  constructor(
+    @InjectEntityManager(geoprocessingConnections.default)
+    private readonly geoprocessingEntityManager: EntityManager,
+  ) {}
+
+  async linkScenarioToCostSurface(
+    scenarioId: string,
+    costSurfaceId: string,
+  ): Promise<void> {
+    await this.geoprocessingEntityManager.transaction(async (em) => {
+      const costsForScenarioPus: {
+        scenariosPuId: string;
+        cost: number;
+      }[] = await em
+        .createQueryBuilder()
+        .select('spd.id', 'scenariosPuId')
+        .addSelect('csp.cost', 'cost')
+        .from(ScenariosPuPaDataGeo, 'spd')
+        .leftJoin(
+          CostSurfacePuDataEntity,
+          'csp',
+          'csp.projects_pu_id = spd.project_pu_id',
+        )
+        .where('spd.scenario_id = :scenarioId', { scenarioId })
+        .andWhere('csp.cost_surface_id = :costSurfaceId', { costSurfaceId })
+        .execute();
+
+      await em.query(
+        ` DELETE FROM scenarios_pu_cost_data spcd
+                       USING scenarios_pu_data spd
+                       WHERE spcd.scenarios_pu_data_id = spd.id and spd.scenario_id = $1`,
+        [scenarioId],
+      );
+
+      await Promise.all(
+        chunk(costsForScenarioPus, CHUNK_SIZE_FOR_BATCH_GEODB_OPERATIONS).map(
+          async (rows) => {
+            await em.insert(
+              ScenariosPuCostDataGeo,
+              rows.map((row) => ({
+                cost: row.cost,
+                scenariosPuDataId: row.scenariosPuId,
+              })),
+            );
+          },
+        ),
+      );
+    });
```

Comment:
since this operation should basically be a copy from one table to another, have you considered doing this in a single SQL query using `UPDATE scenarios_pu_cost_data set cost = cost_surface.cost FROM (SELECT ... ) cost_surface WHERE scenarios_pu_cost_data.scenarios_pu_data_id = spd.scenarios_pu_data_id`? (the `cost_surface` subquery will need to go through a couple of joins to match costs from the `cost_surface_pu_data` table for the desired cost surface to the `scenarios_pu_data_id` values)

Reply:
this will not work for the initial setting of costs, of course, since there won't be any records to update, so that should be handled with an `INSERT`, though this could use the same strategy. After that, the `UPDATE` should be possible as we'll always be dealing with a full set of PUs, and the PUs themselves are guaranteed to be stable (within a scenario, in this case).